File: atlas/examples/typescript/packages/app/src/CatalogDemo.ts
import { CatalogQueryService, ImportCatalog } from '@atlas-example/application';
import { format } from 'date-fns';
import { startCase } from 'lodash-es';
import { RuntimeOutputWriter } from './RuntimeOutputWriter.js';

/**
 * Runs the catalog use cases and renders their results for a terminal user.
 */
export class CatalogDemo {
  /**
   * Creates the delivery workflow from application services and output boundary.
   *
   * @param importCatalog - Imports the configured catalog seed.
   * @param catalogQueries - Retrieves imported catalog items.
   * @param output - Writes intentionally user-facing output.
   */
  public constructor(
    private readonly importCatalog: ImportCatalog,
    private readonly catalogQueries: CatalogQueryService,
    private readonly output: RuntimeOutputWriter,
  ) {}

  /**
   * Imports and presents the complete example catalog.
   */
  public run(): void {
    const report = this.importCatalog.execute();
    const importedAt = format(report.importedAt, 'yyyy-MM-dd');
    this.output.writeLine(
      `Imported ${report.importedCount} catalog items on ${importedAt}.`,
    );
    this.catalogQueries.list().forEach((item) => {
      this.output.writeLine(`${startCase(item.kind)}: ${item.describe()}`);
    });
  }
}

File: atlas/examples/kotlin/app/src/main/kotlin/dev/atlas/example/app/CatalogDemo.kt
package dev.atlas.example.app

import com.google.common.collect.ImmutableList
import org.apache.commons.lang3.StringUtils
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

/**
 * Runs the catalog use cases and renders their results for a terminal user.
 */
class CatalogDemo(
    private val catalogService: ApplicationCatalogService,
    private val output: RuntimeOutputWriter
) {
    /**
     * Imports and presents the complete example catalog.
     */
    fun run() {
        val report = catalogService.importCatalog()
        val importedAt = DateTimeFormatter.ISO_LOCAL_DATE
            .withZone(ZoneOffset.UTC)
            .format(report.importedAt)
        val lines = ImmutableList.builder<String>()
            .add("Imported ${report.importedCount} catalog items on $importedAt.")
            .addAll(
                catalogService.listItems().map { item ->
                    "${StringUtils.capitalize(item.kind.name.lowercase())}: ${item.describe()}"
                }
            )
            .build()
        lines.forEach(output::writeLine)
    }
}
